library/psa_crypto.c

```diff
@@ -6395,4 +6408,163 @@ psa_status_t psa_crypto_init( void )
     return( status );
 }
 
+psa_status_t psa_pake_setup(
+    psa_pake_operation_t *operation,
+    const psa_pake_cipher_suite_t *cipher_suite )
+{
+    return( psa_driver_wrapper_pake_setup( operation, cipher_suite ) );
```

Comment:
I agree that checks from 210-218 can be moved to `psa_pake_setup`, but not check of `operation->alg` since it depends on selected operation context.

Reply:
I agree with Ronald that the core should be responsible for checking the order of PAKE API calls (and the different steps). As a driver implementer, it would be nice to not have to replicate some of logic described around [here](https://github.com/Mbed-TLS/mbedtls/pull/6567/files#diff-086f04b5673d67ec5b33a1115ab83b8257abc0ed6d1a1c86c706f049877e4dc8R101). Similar types of checks will probably be needed for other PAKEs, so adding an alg-dependent stage union could maybe be an option? Maybe this can replace/extend the current `operation->stage` member of `psa_pake_operation_s`?

Reply:
I assume that the order of calls to the pake APIs is done is core already.
- collecting inputs stage:
`psa_pake_setup()` must be called first and it sets `alg` in inputs structure that is required for further API calls to collect inputs: `psa_pake_set_password_key()`, `psa_pake_set_role()`, `psa_pake_set_user()`, `psa_pake_set_peer()`. These functions can be called in any order. Also core verifies if init is not done second time also based on `alg`.
- computation stage
Call to `psa_pake_output()` or `psa_pake_input()` must be performed when all inputs are collected and this is verified before calling `psa_driver_wrapper_pake_setup()`. When `psa_driver_wrapper_pake_setup()` is successfully finished, then the `stage` is changed to `computation`. 
In computation stage `psa_pake_setup()`, `psa_pake_set_password_key()`, `psa_pake_set_role()`, `psa_pake_set_user()`, `psa_pake_set_peer()` can't be called and only sequence of `psa_pake_output()`/`psa_pake_input()/psa_pake_get_implicit_key()` is allowed.

The problem that I see now is with `psa_pake_get_implicit_key()`. There should be check added for the stage(allowed only in computation stage) and it seems core is not able to verify if the implicit key is ready to be read.

Reply:
I agree that those existing checks are good and should be there.  What I'm talking is more related to algorithm-specific state. With J-PAKE as an example, the driver is forced to keep track of the current stage in order to know what to output when being dispatched from `psa_pake_output(&operation, PSA_PAKE_STEP_KEY_SHARE, ...)`. Doing that won't necessarily be difficult (although I don't think it's completely trivial), but my main point is that all drivers would need to do it. Because of that, my opinion is that the core should take that responsibility from the drivers.

Reply:
I see the point now. Current implementation keeps its `stage` in `state/sequence/input_step/output_step` fields.
You suggest to move these fields out of the driver context (`mbedtls_psa_pake_operation_t`) to the psa context (`psa_pake_operation_t`), update these values in core and verify `input`/`output`/`get_implicit_key` calls against current `stage` also in core. 
This make sense, as you mentioned, it will be common part for all drivers. On the other hand this will impose drivers to strictly relay on the core `stages`. Driver entry points for `input`/`output`/`get_implicit_key` will require extra parameter for current `stage`.
I could analyse the code and try to change this, but I think this should be approved first.
cc @mpg @gilles-peskine-arm @ronald-cron-arm

Reply:
Generally speaking, it's the core's responsibility to call drivers in a correct sequence. Drivers are expected to just do the math and not care whether it makes sense. (Drivers do need to validate inputs however, because that can be computationally intensive sometimes, e.g. checking that the public point is on the curve for ECDH. I want to add an explicit list of responsibilities for each mechanism to the driver specification.)

I don't know exactly what this means for PAKE.

Reply:
I will investigate handling pake computation stage by core.